We start our attempt at exploration 7.5 here. We want to compare different numerical schemes in solving ODEs.

2a
____
$x'=x$
The solution to this is $x(t)= e^x + C$. Insert initial condition $x(0)=1$ to discover that $C=0$. Hence, The solution to the differential equation is $x(t)= e^x$. Really appreciate the free gift from the math department.

2b
____
We define the euler's method and find the approximation in the code below

In [4]:
def euler(f,t0,y0,h,n):
    y = y0
    t = t0
    pts = [(t,y)]
    for _ in range(0,n):
        y = y + h * f(t,y)
        t = t + h
        pts.append((t,y))
    return pts

In [5]:
def f(t,y):
    return y

In [6]:
def soln(y):
    return np.exp(y)

In [7]:
#step size= 0.1euler
pts = euler(f,0,1,0.1,10)
print(pts)
#Error
error1 = pts[-1][1]-soln(pts[-1][0])
print(error1)

[(0, 1), (0.1, 1.1), (0.2, 1.2100000000000002), (0.30000000000000004, 1.3310000000000002), (0.4, 1.4641000000000002), (0.5, 1.61051), (0.6, 1.7715610000000002), (0.7, 1.9487171), (0.7999999999999999, 2.1435888100000002), (0.8999999999999999, 2.357947691), (0.9999999999999999, 2.5937424601)]
-0.124539368359045


Notice here that 'pts' is a list of determined $(t_k, x_k)$ for k=1, 2, ... 10 using $\Delta t = 0.1$, starting with $t_0 = 0$ and $x_0 = 1$ as part 2b asked for.

In [8]:
#step size= 0.05euler
pts1 = euler(f,0,1,0.05,20)
print(pts1)
#Error
error2= pts1[-1][1]-soln(pts1[-1][0])
print(error2)

[(0, 1), (0.05, 1.05), (0.1, 1.1025), (0.15000000000000002, 1.1576250000000001), (0.2, 1.2155062500000002), (0.25, 1.2762815625000004), (0.3, 1.3400956406250004), (0.35, 1.4071004226562505), (0.39999999999999997, 1.477455443789063), (0.44999999999999996, 1.5513282159785162), (0.49999999999999994, 1.628894626777442), (0.5499999999999999, 1.7103393581163142), (0.6, 1.7958563260221299), (0.65, 1.8856491423232364), (0.7000000000000001, 1.9799315994393982), (0.7500000000000001, 2.0789281794113683), (0.8000000000000002, 2.182874588381937), (0.8500000000000002, 2.2920183178010336), (0.9000000000000002, 2.406619233691085), (0.9500000000000003, 2.5269501953756395), (1.0000000000000002, 2.6532977051444213)]
-0.06498412331462466


Now as per part 2c, you find the repitition of part 2b with $\Delta t = 0.05$. In this case, 'pts1' is the list containing the values 2c asks for.

In [9]:
#Compute Error Ratio

ratio1= error1/error2
print(ratio1)

1.9164583902452592


So we notice that when we go from stepsize 0.1 to 0.05, our error almost halvens - precisely is 0.52 times the previous error

In [10]:
#step size= 0.01euler
pts2 = euler(f,0,1,0.01,100)
print(pts2)
#Error
error3= pts2[-1][1]-soln(pts2[-1][0])
print(error3)


[(0, 1), (0.01, 1.01), (0.02, 1.0201), (0.03, 1.030301), (0.04, 1.0406040099999998), (0.05, 1.0510100500999997), (0.060000000000000005, 1.0615201506009997), (0.07, 1.0721353521070096), (0.08, 1.0828567056280798), (0.09, 1.0936852726843607), (0.09999999999999999, 1.1046221254112043), (0.10999999999999999, 1.1156683466653163), (0.11999999999999998, 1.1268250301319696), (0.12999999999999998, 1.1380932804332893), (0.13999999999999999, 1.1494742132376221), (0.15, 1.1609689553699982), (0.16, 1.1725786449236981), (0.17, 1.1843044313729352), (0.18000000000000002, 1.1961474756866646), (0.19000000000000003, 1.2081089504435312), (0.20000000000000004, 1.2201900399479664), (0.21000000000000005, 1.2323919403474461), (0.22000000000000006, 1.2447158597509207), (0.23000000000000007, 1.25716301834843), (0.24000000000000007, 1.2697346485319143), (0.25000000000000006, 1.2824319950172334), (0.26000000000000006, 1.2952563149674057), (0.2700000000000001, 1.3082088781170798), (0.2800000000000001, 1.3212909668

2d
____
In the list of 'pts2' you find the repitition of step 2c instead this time using $\Delta t = 0.01$ to approximate $x(1)$. This satisfies task 2d.

In [11]:
ratio2= error2/error3
print(ratio2)

4.825076326006761


So we notice that when we go from stepsize 0.1 to 0.05, our error almost halvens - precisely is 0.52 times the previous error
When we go from 0.01 to 0.05 for stepsize our error is almost $\frac{1}{4.8}$ th 

2e
_____
Now we try to implement improved euler method with the recursion law given in the text

In [12]:
def improved_euler(f, t0, y0, h, n):
    y = y0
    t = t0
    pts = [(t,y)]
    
    for _ in range(0, n):
        # Predictor step (using Euler method)
        k1 = f(t, y)
        y_predict = y + h * k1
        
        # Corrector step
        k2 = f(t + h, y_predict)
        y = y + h * (k1 + k2) / 2  # Average of slopes
        
        t = t + h
        pts.append((t,y))
    
    return pts


In [13]:
#step size= 0.1imeuler
pts = improved_euler(f,0,1,0.1,10)
print(pts)
#Error
error4= pts[-1][1]-soln(pts[-1][0])
print(error4)

[(0, 1), (0.1, 1.105), (0.2, 1.221025), (0.30000000000000004, 1.349232625), (0.4, 1.490902050625), (0.5, 1.647446765940625), (0.6, 1.8204286763643904), (0.7, 2.0115736873826515), (0.7999999999999999, 2.22278892455783), (0.8999999999999999, 2.456181761636402), (0.9999999999999999, 2.714080846608224)]
-0.004200981850821073


Here we print 'pts' to approximate using improved euler's method using $\Delta t = 0.1$ just like before

In [14]:
#step size= 0.05imeuler
pts = improved_euler(f,0,1,0.05,20)
print(pts)
#Error
error5= pts[-1][1]-soln(pts[-1][0])
print(error5)



[(0, 1), (0.05, 1.05125), (0.1, 1.1051265625), (0.15000000000000002, 1.161764298828125), (0.2, 1.2213047191430664), (0.25, 1.2838965859991485), (0.3, 1.3496962860316049), (0.35, 1.4188682206907246), (0.39999999999999997, 1.4915852170011243), (0.44999999999999996, 1.568028959372432), (0.49999999999999994, 1.648390443540269), (0.5499999999999999, 1.7328704537717077), (0.6, 1.8216800645275077), (0.65, 1.9150411678345425), (0.7000000000000001, 2.0131870276860626), (0.7500000000000001, 2.1163628628549733), (0.8000000000000002, 2.2248264595762905), (0.8500000000000002, 2.3388488156295755), (0.9000000000000002, 2.4587148174305913), (0.9500000000000003, 2.584723951823909), (1.0000000000000002, 2.7171910543548843)]
-0.0010907741041616958


Here we print 'pts' to approximate using improved euler's method using $\Delta t = 0.05$ just like before

In [15]:
ratio3= error4/error5
print(ratio3)

3.8513765909850766


When we go from stepsize 0.1 to 0.05, our error is $\frac{1}{3.85}$ th

In [16]:
#step size= 0.01imeuler
pts = improved_euler(f,0,1,0.01,100)
print(pts)
#Error
error6= pts[-1][1]-soln(pts[-1][0])
print(error6)

[(0, 1), (0.01, 1.01005), (0.02, 1.0202010025), (0.03, 1.0304540225751249), (0.04, 1.0408100855020048), (0.05, 1.0512702268613), (0.060000000000000005, 1.0618354926412559), (0.07, 1.0725069393423006), (0.08, 1.0832856340826906), (0.09, 1.0941726547052217), (0.09999999999999999, 1.1051690898850093), (0.10999999999999999, 1.1162760392383537), (0.11999999999999998, 1.127494613432699), (0.12999999999999998, 1.1388259342976976), (0.13999999999999999, 1.1502711349373895), (0.15, 1.1618313598435104), (0.16, 1.1735077650099377), (0.17, 1.1853015180482875), (0.18000000000000002, 1.197213798304673), (0.19000000000000003, 1.209245796977635), (0.20000000000000004, 1.2213987172372602), (0.21000000000000005, 1.2336737743454946), (0.22000000000000006, 1.2460721957776668), (0.23000000000000007, 1.2585952213452323), (0.24000000000000007, 1.271244103319752), (0.25000000000000006, 1.2840201065581154), (0.26000000000000006, 1.2969245086290244), (0.2700000000000001, 1.3099585999407461), (0.2800000000000001

Here we print 'pts' to approximate using improved euler's method using $\Delta t = 0.01$ just like before

In [17]:
ratio4= error5/error6
print(ratio4)

24.257807054609312


When we go from stepsize 0.05 to 0.01, our error is $\frac{1}{24.25}$ th

2f
______

Now we implement the Runge-Kutta method.

In [18]:
def runge_kutta(f,t0,y0,h,n):
    y = y0
    t = t0
    pts = [(t,y)]
    for _ in range(0,n):
        k1 = f(t,y)
        k2 = f(t+h/2,y+k1*h/2)
        k3 = f(t+h/2,y+k2*h/2)
        k4 = f(t+h,y+k3*h)
        y = y + (k1+2*k2+2*k3+k4)*h/6
        t = t + h
        pts.append((t,y))
    return pts

In [19]:
#step size= 0.1rk
pts = runge_kutta(f,0,1,0.1,10)
print(pts)
#Error
error7= pts[-1][1]-soln(pts[-1][0])
print(error7)

[(0, 1), (0.1, 1.1051708333333334), (0.2, 1.2214025708506946), (0.30000000000000004, 1.3498584970625378), (0.4, 1.4918242400806858), (0.5, 1.6487206385968383), (0.6, 1.8221179620919332), (0.7, 2.0137516265967768), (0.7999999999999999, 2.2255395632923154), (0.8999999999999999, 2.459601413780071), (0.9999999999999999, 2.718279744135166)]
-2.0843238792700447e-06


Here we print 'pts' to approximate using runge- kutta method using $\Delta t = 0.1$ just like before for 2f.

In [20]:
#step size= 0.05rk
pts = runge_kutta(f,0,1,0.05,20)
print(pts)
#Error
error8= pts[-1][1]-soln(pts[-1][0])
print(error8)

[(0, 1), (0.05, 1.05127109375), (0.1, 1.1051709125543214), (0.15000000000000002, 1.1618342340216672), (0.2, 1.2214027459561516), (0.25, 1.284025400650577), (0.3, 1.3498587873447139), (0.35, 1.419067523779926), (0.39999999999999997, 1.4918246678292268), (0.44999999999999996, 1.5683121502320618), (0.49999999999999994, 1.648721229515874), (0.5499999999999999, 1.7332529702419976), (0.6, 1.8221187457717412), (0.65, 1.9155407668098365), (0.7000000000000001, 2.0137526370468906), (0.7500000000000001, 2.1169999372902315), (0.8000000000000002, 2.225540839543783), (0.8500000000000002, 2.339646752572486), (0.9000000000000002, 2.4596030005655134), (0.9500000000000003, 2.585709536595289), (1.0000000000000002, 2.718281692656335)]
-1.3580271085444906e-07


Here we print 'pts' to approximate using runge- kutta method using $\Delta t = 0.05$ just like before for 2f.

In [21]:
ratio5= error7/error8 
print(ratio5)

15.348175792337358


In [22]:
#step size= 0.01rk
pts = runge_kutta(f,0,1,0.01,100)
print(pts)
#Error
error9= pts[-1][1]-soln(pts[-1][0])
print(error9)


[(0, 1), (0.01, 1.0100501670833333), (0.02, 1.0202013400250696), (0.03, 1.030454533950962), (0.04, 1.0408107741889476), (0.05, 1.05127109637168), (0.060000000000000005, 1.0618365465400945), (0.07, 1.0725081812480122), (0.08, 1.0832870676677966), (0.09, 1.0941742836970723), (0.09999999999999999, 1.1051709180665144), (0.10999999999999999, 1.1162780704487236), (0.11999999999999998, 1.1274968515681942), (0.12999999999999998, 1.138828383312387), (0.13999999999999999, 1.1502737988439187), (0.15, 1.1618342427138806), (0.16, 1.1735108709762931), (0.17, 1.185304851303713), (0.18000000000000002, 1.1972173631040008), (0.19000000000000003, 1.2092495976382638), (0.20000000000000004, 1.221402758139982), (0.21000000000000005, 1.233678059935333), (0.22000000000000006, 1.2460767305647256), (0.23000000000000007, 1.2586000099055548), (0.24000000000000007, 1.2712491502961907), (0.25000000000000006, 1.2840254166612128), (0.26000000000000006, 1.2969300866379048), (0.2700000000000001, 1.3099644507040178), (0

Here we print 'pts' to approximate using runge- kutta method using $\Delta t = 0.01$ just like before for 2f.

In [23]:
ratio6= error8/error9
print(ratio6)

604.5268646300985


When we go from stepsize 0.05 to 0.01, our error is $\frac{1}{604.52}$ th

2g
____

Throughout this notebook till now, you have come across error computations already. They are printed as 'error1', 'error2'....,'error9'. We also used the given value of $e$ to calculate the errors. This satisfies the completion of 2g.

3
_____

Now we do the second ODE $x' = 2t(1+x^2)$. Taking the initial condition $x(0)=0$, we get the analytical solution to be $x(t)= tan(t^2)$ 

In [24]:
import numpy as np

def euler(f,t0,y0,h,n):
    y = y0
    t = t0
    pts = [(t,y)]
    for _ in range(0,n):
        y = y + h * f(t,y)
        t = t + h
        pts.append((t,y))
    return pts

def f(t,y):
    return 2*t*(1+y**2)

def soln(t):
    return np.tan(t**2)

# List of step sizes and corresponding n values
step_sizes = [0.1, 0.05, 0.01]
n_values = [10, 20, 100]

errors = []

# Loop to compute and print errors for each step size
for h, n in zip(step_sizes, n_values):
    pts = euler(f,0,0,h,n)
    print(f"Points for step size {h}:")
    print(pts)
    
    error = pts[-1][1] - soln(pts[-1][0])
    errors.append(error)
    
    print(f"Error for step size {h}:", error)
    print()

# Compute and print ratios
for i in range(len(errors)-1):
    ratio = errors[i] / errors[i+1]
    print(f"Ratio {i+1}:", ratio)

Points for step size 0.1:
[(0, 0), (0.1, 0.0), (0.2, 0.020000000000000004), (0.30000000000000004, 0.06001600000000001), (0.4, 0.12023211521536002), (0.5, 0.2013885761376928), (0.6, 0.3054443119975695), (0.7, 0.43663985932536975), (0.7999999999999999, 0.6033314706706048), (0.8999999999999999, 0.8215728888308536), (0.9999999999999999, 1.123069650929991)]
Error for step size 0.1: -0.4343380737249105

Points for step size 0.05:
[(0, 0), (0.05, 0.0), (0.1, 0.005000000000000001), (0.15000000000000002, 0.015000250000000001), (0.2, 0.030003625112500938), (0.25, 0.05002162946289877), (0.3, 0.07508418354825186), (0.35, 0.10525331258682509), (0.39999999999999997, 0.14064105168019259), (0.44999999999999996, 0.18143224789690102), (0.49999999999999994, 0.22791354262286254), (0.5499999999999999, 0.28051077176840766), (0.6, 0.3398385178877036), (0.65, 0.4067679309821102), (0.7000000000000001, 0.4825228407110156), (0.7500000000000001, 0.5688208211375636), (0.8000000000000002, 0.6680876056295345), (0.85

For convenience, approximations for $\Delta t = 0.1,0.05, 0.01$, their corresponding errors, error ratios using euler method are all printed above. 

In [25]:
import numpy as np

def improved_euler(f, t0, y0, h, n):
    y = y0
    t = t0
    pts = [(t,y)]
    
    for _ in range(0, n):
        # Predictor step (using Euler method)
        k1 = f(t, y)
        y_predict = y + h * k1
        
        # Corrector step
        k2 = f(t + h, y_predict)
        y = y + h * (k1 + k2) / 2  # Average of slopes
        
        t = t + h
        pts.append((t,y))
    
    return pts
def f(t,y):
    return 2*t*(1+y**2)

def soln(t):
    return np.tan(t**2)

# List of step sizes and corresponding n values
step_sizes = [0.1, 0.05, 0.01]
n_values = [10, 20, 100]

errors = []

# Loop to compute and print errors for each step size
for h, n in zip(step_sizes, n_values):
    pts = improved_euler(f,0,0,h,n)
    print(f"Points for step size {h}:")
    print(pts)
    
    error = pts[-1][1] - soln(pts[-1][0])
    errors.append(error)
    
    print(f"Error for step size {h}:", error)
    print()

# Compute and print ratios
for i in range(len(errors)-1):
    ratio = errors[i] / errors[i+1]
    print(f"Ratio {i+1}:", ratio)

Points for step size 0.1:
[(0, 0), (0.1, 0.010000000000000002), (0.2, 0.04001900240008001), (0.30000000000000004, 0.0902434317215935), (0.4, 0.16139655423946483), (0.5, 0.25540264491845655), (0.6, 0.3765235830364665), (0.7, 0.5334901280365955), (0.7999999999999999, 0.7441207816820764), (0.8999999999999999, 1.0471115706672351), (0.9999999999999999, 1.5387032414988446)]
Error for step size 0.1: -0.018704483156056773

Points for step size 0.05:
[(0, 0), (0.05, 0.0025000000000000005), (0.1, 0.010000296877343755), (0.15000000000000002, 0.022503797296125568), (0.2, 0.04002166649932644), (0.25, 0.06258276442503065), (0.3, 0.09024704047839369), (0.35, 0.12312327640908445), (0.39999999999999997, 0.1613919857836359), (0.44999999999999996, 0.2053349714564435), (0.49999999999999994, 0.25537415469141783), (0.5499999999999999, 0.31212408084820775), (0.6, 0.37646546152274146), (0.65, 0.44965216573600764), (0.7000000000000001, 0.5334730852594245), (0.7500000000000001, 0.6305071414496614), (0.800000000

For convenience, approximations for $\Delta t = 0.1,0.05, 0.01$, their corresponding errors, error ratios using improved euler method are all printed above. 

In [26]:
def runge_kutta(f,t0,y0,h,n):
    y = y0
    t = t0
    pts = [(t,y)]
    for _ in range(0,n):
        k1 = f(t,y)
        k2 = f(t+h/2,y+k1*h/2)
        k3 = f(t+h/2,y+k2*h/2)
        k4 = f(t+h,y+k3*h)
        y = y + (k1+2*k2+2*k3+k4)*h/6
        t = t + h
        pts.append((t,y))
    return pts


def f(t,y):
    return 2*t*(1+y**2)

def soln(t):
    return np.tan(t**2)

# List of step sizes and corresponding n values
step_sizes = [0.1, 0.05, 0.01]
n_values = [10, 20, 100]

errors = []

# Loop to compute and print errors for each step size
for h, n in zip(step_sizes, n_values):
    pts = runge_kutta(f,0,0,h,n)
    print(f"Points for step size {h}:")
    print(pts)
    
    error = pts[-1][1] - soln(pts[-1][0])
    errors.append(error)
    
    print(f"Error for step size {h}:", error)
    print()

# Compute and print ratios
for i in range(len(errors)-1):
    ratio = errors[i] / errors[i+1]
    print(f"Ratio {i+1}:", ratio)

Points for step size 0.1:
[(0, 0), (0.1, 0.010000416683333542), (0.2, 0.04002168071683565), (0.30000000000000004, 0.09024454675887025), (0.4, 0.16138084198410918), (0.5, 0.2553442224512059), (0.6, 0.37640664469928414), (0.7, 0.5333948015299372), (0.7999999999999999, 0.744556949309006), (0.8999999999999999, 1.0504828722528272), (0.9999999999999999, 1.5574275302059482)]
Error for step size 0.1: 1.9805551046747638e-05

Points for step size 0.05:
[(0, 0), (0.05, 0.002500006510432943), (0.1, 0.010000338555379752), (0.15000000000000002, 0.022503809369423), (0.2, 0.04002136787407374), (0.25, 0.06258154031076518), (0.3, 0.09024383740411121), (0.35, 0.12311652121094781), (0.39999999999999997, 0.16137954798961002), (0.44999999999999996, 0.20531420292750188), (0.49999999999999994, 0.2553420685279132), (0.5499999999999999, 0.31207778814408316), (0.6, 0.37640310074835215), (0.65, 0.4495748042341768), (0.7000000000000001, 0.5333886049684969), (0.7500000000000001, 0.6304383306476673), (0.800000000000

For convenience, approximations for $\Delta t = 0.1,0.05, 0.01$, their corresponding errors, error ratios using runge-kutta method are all printed above. 

Throughout the codes snippet above, you find a holistic exploration of $x' = 2t(1+x^2)$ satisfying the steps of 3.